mirai-core-utils/src/commonMain/kotlin/Resources.kt

```diff
@@ -9,8 +9,93 @@
 
 package net.mamoe.mirai.utils
 
+import java.util.concurrent.atomic.AtomicInteger
+
 
 @TestOnly
 public fun readResource(url: String): String =
     Thread.currentThread().contextClassLoader.getResourceAsStream(url)?.readBytes()?.decodeToString()
         ?: error("Could not find resource '$url'")
+
+public class ResourceAccessLock {
+    // -2: Locked
+    // -1: Uninitialized
+    // 0: Initialized
+    // >0: Using
+    private val count = AtomicInteger(-1)
+
+    /**
+     * ```
+     * if (res.lock.tryToDispose()) {
+     *  res.internal.close()
+     * }
+     * ```
+     */
+    public fun tryToDispose(): Boolean {
+        return count.compareAndSet(0, -1)
+    }
+
+    /**
+     * ```
+     * if (res.lock.tryToInit()) {
+     *  res.internalRes = download()
+     * }
+     * ```
+     */
+    public fun tryToInit(): Boolean {
+        return count.compareAndSet(-1, 0)
+    }
+
+    public fun tryToUse(): Boolean {
+        val c = count
+        while (true) {
+            val v = c.get()
+            if (v < 0) return false
+            if (c.compareAndSet(v, v + 1)) return true
+        }
+    }
+
+    public fun lockIfNotUsing(): Boolean {
+        val count = this.count
+        while (true) {
+            val value = count.get()
+            if (value != 0) return false
+            if (count.compareAndSet(0, -2)) return true
+        }
+    }
+
+    public fun release() {
+        count.getAndDecrement()
+    }
+
+    public fun unlock() {
+        count.compareAndSet(-2, 0)
+    }
+
+    public fun setToInit() {
+        count.set(0)
+    }
+
+    public fun setToLocked() {
```

Comment:
```suggestion
    public fun setLocked() {
```

'To' is redundant. Removing for clarity. Same for the other 3 'To's.